### ISSUE TITLE
[REPRODUCTION] Cosmos3-Nano on LIBERO-10: notes to help the next reproduction

### ISSUE THREAD

**Issue by daniel-gieseler** [NVIDIA/cosmos-framework#50]
We reproduced Cosmos3-Nano on LIBERO-10 and matched Table 20's 97.4% at checkpoint 2000.

We're working towards fine-tuning Cosmos3 as a robot policy for the Unitree G1 humanoid. LIBERO-10 is a good warm-up for that: like the G1, it's post-training onto an out-of-distribution embodiment.

A few things weren't obvious from the shipped scripts, so we're writing them down for the next person.

## 1. Complementing the repo

Things that are missing or won't load as shipped:

- The LIBERO recipe isn't on `main` (only the DROID config + server script are) — we found it on
  `mharrim-nv-patch-1`.
- We had to reconstruct two imported files even there: `libero_pose_utils` and
  `cosmos3_action_lerobot._patch_decoder_cache`. 

## 2. Training

The shipped config is fine, just one thing to know:

- The config trains on all four suites by default. That gives libero_10 only ~1 pass in 2000 steps
  → ~82%. The same 2000 steps on libero_10 alone (~2.7 passes) → the 97% in the paper.

## 3. Writing the eval loop — the repo has none

The repo has no closed-loop LIBERO eval client (just the inference server), so we wrote our own.
Conventions we had to match - or else the score tanks:

- **Image**: the sim frames are rotated 180° vs training, so rotate them back with `img[::-1,::-1]`.
- **Gripper**: the model outputs `[0,1]` but robosuite wants `[-1,1]` (negative = open). Convert
  with `1 - 2·g`; pass it through raw and the gripper never opens.
- **Server**: the server's normalization flags already exist — just remember to use them
  (`--action-normalization quantile_rot` + the libero rot6d stats file), or actions come out the
  wrong scale.

Hope this saves someone the digging.

**Comment by lfengad** [comment]
@daniel-gieseler  Thanks for the contribution! We would see to incorporate that into the framework also! THX a lot!

**Comment by gjhnb18** [comment]
Thanks for sharing your detailed notes, this is extremely helpful!
We followed the finding to train on LIBERO-10 only and evaluated each task with 50 episodes. Our per-task success rates are listed below:
task_0: 27/50 = 0.54
task_1: 50/50 = 1.0
task_2: 45/50 = 0.9
task_3: 49/50 = 0.98
task_4: 49/50 = 0.98
task_5: 46/50 = 0.92
task_6: 49/50 = 0.98
task_7: 49/50 = 0.98
task_8: 50/50 = 1.0
task_9: 48/50 = 0.96
We launched training via examples/launch_sft_action_policy_libero.sh on 2×8 H20 GPUs, 2k steps.
Overall performance is close to the reported 97.4%, but task_0 is a clear outlier with much lower success rate. Do you have any insight on whether you also observed weak performance on task_0, or any extra tuning tricks for this specific task?

Another potential factor we are investigating is the maximum timestep limit in our closed-loop evaluator. Currently we cap each rollout at 520 steps. It’s possible that task_0 requires longer motion planning and many trials terminate early due to hitting this threshold, while other easier tasks finish well within the limit. Could you share what episode horizon you used in your evaluation loop?

The training detail is following：

<img width="814" height="475" alt="Image" src="https://github.com/user-attachments/assets/1b52a134-0bca-48f3-a9c3-ff60c587e440" />

**Comment by fwd4** [comment]
Thanks for reporting back your findings! @gjhnb18 

> Do you have any insight on whether you also observed weak performance on task_0, or any extra tuning tricks for this specific task?

We do find task_0 to be an outlier but not with such huge discrepancy in success rate, in our evaluation we find it to drop down to ~80%. There is no other tricks applied for this specific task.

> Could you share what episode horizon you used in your evaluation loop?

We are using the same episode horizon as you do, nothing special is changed here.
